Classified listings refactor part 3

citizen428 opened the pull request:
## What type of PR is this? (check all applicable)

- [X] Refactor

## Description

This is step 3 of refactoring `ClassifiedListing` from using hardcoded categories to DB managed categories. 

1. Ignore column
2. Move view oriented class methods into helper.

After this, there will be one more PR to actually drop the column from the table.

## Related Tickets & Documents

#6740

## Mobile & Desktop Screenshots/Recordings (if there are UI changes)

## Added tests?

- [x] yes for new things, refactored for old

## Added to documentation?

- [X] no documentation needed

## What gif best describes this PR or how it makes you feel?

![image](https://user-images.githubusercontent.com/47985/80188041-d4860700-863a-11ea-8c5e-711ed05c81d6.png)

citizen428 commented:
This is not actually ready for review, but Travis seems to have gotten completely stuck on this PR :sad: Neither @rhymes attempt not my last commit seem to actually have triggered a build.

citizen428 commented:
In the end, I stopped myself from refactoring everything I wanted to refactor, this took long enough. But quite a lot of code good cleaned up/made at least a little better.

This PR also adds an Administrate dashboard for categories, so admins can manage them.

The main outstanding issue (which I'll tackle in another PR) is this method in `app/controllers/social_previews_controller.rb`:

```ruby
  # TODO: [thepracticaldev/oss] don't hardcode this
  def define_categories
    cat_info = {
      "collabs": ["Collaborators Wanted", "#5AE8D9"],
      "cfp": ["Call For Proposal", "#f58f8d"],
      "forhire": ["Available For Hire", "#b78cf4"],
      "education": ["Education", "#5AABE8"],
      "jobs": ["Now Hiring", "#53c3ad"],
      "mentors": ["Offering Mentorship", "#A69EE8"],
      "mentees": ["Looking For Mentorship", "#88aedb"],
      "forsale": ["Stuff For Sale", "#d0adfb"],
      "events": ["Upcoming Event", "#f8b3d0"],
      "misc": ["Miscellaneous", "#6393FF"],
      "products": ["Products & Tools", "#5AE8D9"]
    }
    @category = cat_info[@listing.category.to_sym][0]
    @cat_color = cat_info[@listing.category.to_sym][1]
  end
```

Not sure what the best course of action is here. Should we add the color as a column to the listing model and let admins manage it through there? Would we then also need another column for the short description?

rhymes commented on app/models/classified_listing.rb:
```diff
@@ -1,15 +1,16 @@
 class ClassifiedListing < ApplicationRecord
+  self.ignored_columns = ["category"]
```
🔥

rhymes commented on app/views/classified_listings/index.html.erb:
```diff
@@ -28,25 +28,19 @@
-      }
-    ) %>"
-    data-allcategories="<%= ClassifiedListing.categories_for_display.to_json %>"
+    data-category="<%= params[:category] %>" data-listings="<%= @listings_json %>"
```
yes! :D

rhymes commented on app/views/classified_listings/index.html.erb:
```diff
@@ -28,25 +28,19 @@
     <div class="classified-filters">
     <div class="classified-filters-categories">
-      <a href="/listings" class="<%= "selected" if params[:category].blank? %> data-no-instant="true">all</a><% ClassifiedListing.categories_for_display.each do |cat| %><a href="/listings/<%= cat[:slug] %>" class="<%= "selected" if params[:category] == cat[:slug] %> data-no-instant="true"><%= cat[:name] %></a><% end %><a href='/listings/new' class='classified-create-link'>Create a Listing</a>
+      <a href="/listings" class="<%= "selected" if params[:category].blank? %> data-no-instant="true">all</a>
```
these harcoded paths could use the helpers as well if you want to

rhymes commented on db/schema.rb:
```diff
@@ -308,6 +308,7 @@
     t.string "slug"
     t.string "status", default: "active"
     t.datetime "updated_at", null: false
+    t.index ["slug"], name: "index_chat_channels_on_slug", unique: true
```
this is definitely gone, it has been removed on master I think

rhymes commented on db/schema.rb:
```diff
@@ -1216,9 +1217,9 @@
-    t.string "youtube_url"
     t.boolean "welcome_notifications", default: true, null: false
     t.datetime "workshop_expiration"
+    t.string "youtube_url"
```
I think we can keep this change in as it keeps popping up, there was a problem of ordering of fields I think

vaidehijoshi replied:
+1 for this!!! I was just lamenting today why this keeps popping up in my diff and @rhymes pointed me here! Just wanted to say thank you for fixing this, @citizen428

rhymes commented on db/seeds.rb:
```diff
@@ -371,6 +371,49 @@
 
 ##############################################################################
 
+counter += 1
```
this change has to go, see #7512

mstruve commented on app/controllers/api/v0/classified_listings_controller.rb:
```diff
@@ -55,7 +55,7 @@ def update
       ATTRIBUTES_FOR_SERIALIZATION = %i[
         id user_id organization_id title slug body_markdown cached_tag_list
-        category classified_listing_category_id category processed_html published
+        classified_listing_category_id processed_html published
```
Since we are removing a field here are we concerned about the fact that this change is not really backwards compatible? 

Also, would it be helpful to keep it so the user doesn't need to go look up the classified_listing_category_id if they want to know what the category name is?

citizen428 replied:
At one point we kinda said that as long as the API is labeled v0, we're ok with minor breakage. That said, this list just includes the attributes we select from the DB. The serializer adds it back it (it's a model method delegating to the association):

```ruby
json.type_of "classified_listing"

json.extract!(
  listing,
  :id,
  :title,
  :slug,
  :body_markdown,
  :category,
  :classified_listing_category_id,
  :processed_html,
  :published,
)

json.tag_list listing.cached_tag_list
json.tags     listing.tag_list
```

mstruve replied:
OH thanks for the clarification didn't realize this was a db retrieval list!

mstruve commented on app/controllers/concerns/classified_listings_toolkit.rb:
```diff
@@ -4,33 +4,22 @@ module ClassifiedListingsToolkit
-    @classified_listing.save
+    # [thepracticaldev/oss] Not entirely sure what the intention behind the
+    # original code was, but at least this is more compact.
+    filtered_params = listing_params.reject { |_k, v| v.nil? }
```
🙌

mstruve commented on app/liquid_tags/classified_listing_tag.rb:
```diff
@@ -2,8 +2,8 @@ class ClassifiedListingTag < LiquidTagBase
   def initialize(_tag_name, slug_path_url, _tokens)
-    striped_path = ActionController::Base.helpers.strip_tags(slug_path_url).strip
-    @listing = get_listing(striped_path)
+    stripped_path = ActionController::Base.helpers.strip_tags(slug_path_url).strip
```
😂

citizen428 commented:
Thanks for the feedback @rhymes & @mstruve.

* Updated `seeds.rb`. both typo and merge/rebase problems.
* Updated `schema.rb` and removed wrongly re-added column.
